### ISSUE TITLE
Azure Issue .NET Core 3.1: System.ComponentModel.Win32Exception (126): The specified module could not be found.

### ISSUE THREAD

**Issue by upsilondynamics** [microsoft/ClearScript#171]
Issue in Azure web app (Works 100% locally) on only one of my Web projects, but not the other, no clue what the issue is - @ClearScriptLib whats the problem here that would make it work on one Web API project, but not the other? (.NET Core 3.1 FYI, 64-BIT)

This was working before and now it isn't for some reason, like I said I have 2 web apps; one works fine and this one doesn't. Please assist:

```
  4/1/2020 2:45:30 PM	Method GenerateFlow: System.ComponentModel.Win32Exception (126): The specified module could not be found.

     at Microsoft.ClearScript.Util.IJW.IJWHostLibrary.Load()

     at Microsoft.ClearScript.V8.V8Proxy.LoadAssemblyWithIJWHostLibrary()

     at Microsoft.ClearScript.V8.V8Proxy.GetAssembly()

     at Microsoft.ClearScript.V8.V8Proxy.GetImplType(Type type)

     at Microsoft.ClearScript.V8.V8Proxy.CreateImpl[T](Object[] args)

     at Microsoft.ClearScript.V8.V8IsolateProxy.Create(String name, V8RuntimeConstraints constraints, V8RuntimeFlags flags, Int32 debugPort)

     at Microsoft.ClearScript.V8.V8Runtime..ctor(String name, V8RuntimeConstraints constraints, V8RuntimeFlags flags, Int32 debugPort)

     at Microsoft.ClearScript.V8.V8Runtime..ctor(String name, V8RuntimeConstraints constraints, V8RuntimeFlags flags)

     at Microsoft.ClearScript.V8.V8Runtime..ctor(String name, V8RuntimeConstraints constraints)

     at Microsoft.ClearScript.V8.V8ScriptEngine..ctor(V8Runtime runtime, String name, V8RuntimeConstraints constraints, V8ScriptEngineFlags flags, Int32 debugPort)

     at Microsoft.ClearScript.V8.V8ScriptEngine..ctor(String name, V8RuntimeConstraints constraints, V8ScriptEngineFlags flags, Int32 debugPort)

     at Microsoft.ClearScript.V8.V8ScriptEngine..ctor(String name, V8RuntimeConstraints constraints, V8ScriptEngineFlags flags)

     at Microsoft.ClearScript.V8.V8ScriptEngine..ctor(String name, V8RuntimeConstraints constraints)

     at Microsoft.ClearScript.V8.V8ScriptEngine..ctor(String name)
```

Segment of code failing:

```
  _engine = new V8ScriptEngine("JFlowScriptEngine")
            {
                AllowReflection = false,
                MaxRuntimeHeapSize = (UIntPtr) MaxRuntimeHeapSize,
                MaxRuntimeStackUsage = (UIntPtr) MaxRuntimeStackUsage
            };
```

**Comment by ClearScriptLib** [comment]
Hello @upsilondynamics,

At first glance, this would appear to be a duplicate of Issue #160. Can you try the workaround of using LocalSystem or a local/domain account for the application pool?

We have a fix pending for that issue, and the next ClearScript release should be out in a couple of weeks, as soon as V8 8.1 reaches the stable channel.

Good luck!

**Comment by upsilondynamics** [comment]
@ClearScriptLib  As this is an Azure hosted WebApp I do not think I can change the IIS App pool identity.

**Comment by ClearScriptLib** [comment]
@upsilondynamics Are you still using one of the Azure workarounds listed [here](https://github.com/microsoft/ClearScript/issues/166#issuecomment-583903196)?

**Comment by upsilondynamics** [comment]
@ClearScriptLib Indeed I am, I am manually placing the **vcruntime140_1.dll** in the webroot and it fixed the other issue, but this seems to be a different issue.

**Comment by upsilondynamics** [comment]
I started looking at this class here: https://github.com/microsoft/ClearScript/blob/master/NetCore/ClearScript/Util/IJW/IJWHostLibrary.cs

I am assuming its failing at line 19 when it tries to find that dll.  Where does this DLL come from? Is it possible it doesn't exist on the host my webapp is hosted on?

**Microsoft.ClearScript.Util.IJW.ijwhost-64.dll**

Thanks

**Comment by ClearScriptLib** [comment]
>Where does this DLL come from?

It's generated whenever a .NET Core mixed mode-assembly is linked, becoming a dependency of that assembly. Unfortunately it's native, so it must be loaded from the file system, and it must have the same hard-coded name (ijwhost.dll) on both 32-bit and 64-bit architectures.

The naming requirement prevents the side-by-side installation of the 32-bit and 64-bit variants, so instead we ship them as embedded resources within the .NET Core version of ClearScript.dll (which is architecture-neutral). On first run the appropriate ijwhost.dll variant is extracted into the file system and loaded explicitly from there.

Most likely you're running into one of two issues. Either ClearScript is unable to copy ijwhost.dll to the file system, or the Azure environment is missing some system dependency that's usually there. Since we can't reproduce this issue, it's difficult to tell exactly what's happening in your case.

**Comment by upsilondynamics** [comment]
Okay let's figure this out, this is the directory it is trying to get the file from:

**C:\DWASFiles\Sites\#1flowdynamics__048e\LocalAppData\Microsoft\ClearScript\6.0.0\x64**

I managed to get this because I reverse engineered how your code is pulling the path back:

```
        public static string GetLocalDataRootPath()
        {
            return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Microsoft", "ClearScript", ClearScriptVersion.Triad, Environment.Is64BitProcess ? "x64" : "x86");
        }
```

On my  local, I can see the **ijwhost.dll** file there; but for some reason I don't think on Azure this path is available or it cannot place the file there.  Kudo console won't even allow me to access the C drive to see if this path or file even exists since I notice the code logic will check there first, and if it doesn't exist it tries to extract the DLL from the resource (like you said, which is fine) and put the file there - this isn;t placing nice with the Azure webapp environment.

Can we make a small change (hot fix) and get me a new DLL where it checks the local path as well for this file, that way I can simply upload it to my webroot to get around this?

**Comment by upsilondynamics** [comment]
Proposed solution, seems to have fixed the issue for me - rebuilt a custom version of ClearScript.dll that works, I also am including manually that iwhost.dll file with my deployment, but if it doesn't exist - the code below should create it properly. 

```
using System;
using System.ComponentModel;
using System.IO;
using Microsoft.ClearScript.Util.COM;
using Microsoft.ClearScript.V8;

namespace Microsoft.ClearScript.Util.IJW
{
    internal static class IJWHostLibrary
    {
        private static string GetLocalMappedPath(string path)
        {
            //Prepare web server characters to localized path
            return System.IO.Path.GetFullPath(path).Replace("~\\", "");
        }

        public static IntPtr Load()
        {
            var dirPath = GetLocalMappedPath("~");
            var filePath = Path.Combine(dirPath, "ijwhost.dll");

            if (!File.Exists(filePath))
            {
                var suffix = Environment.Is64BitProcess ? "64" : "32";
                using (var resourceStream = typeof(V8Proxy).Assembly.GetManifestResourceStream("Microsoft.ClearScript.Util.IJW.ijwhost-" + suffix + ".dll"))
                {
                    if (resourceStream == null)
                    {
                        throw new TypeLoadException("Could not access embedded mixed assembly support library");
                    }

                    Directory.CreateDirectory(dirPath);

                    Stream fileStream = null;
                    try
                    {
                        fileStream = new FileStream(filePath, FileMode.CreateNew, FileAccess.Write);
                    }
                    catch (IOException exception)
                    {
                        if (exception.HResult != HResult.WIN32_E_FILEEXISTS)
                        {
                            throw;
                        }
                    }

                    if (fileStream != null)
                    {
                        using (fileStream)
                        {
                            resourceStream.CopyTo(fileStream);
                        }
                    }
                }
            }

            var hLibrary =  NativeMethods.LoadLibraryW(filePath);
            if (hLibrary != IntPtr.Zero)
            {
                return hLibrary;
            }

            throw new Win32Exception();
        }
    }
}

```

**Comment by ClearScriptLib** [comment]
Hi @upsilondynamics,

Glad to hear that you've found a solution! It shows that no system dependencies are missing, but we still don't know what's going on with your local data folder.

>On my local, I can see the ijwhost.dll file there; but for some reason I don't think on Azure this path is available or it cannot place the file there. Kudo console won't even allow me to access the C drive to see if this path or file even exists since I notice the code logic will check there first, and if it doesn't exist it tries to extract the DLL from the resource (like you said, which is fine) and put the file there - this isn't placing nice with the Azure webapp environment.

Our understanding is that, despite having no access to the C drive in general, applications _should_ have access to their local data folders no matter where they are.

Issue #160 shows that in some cases ASP.NET Core applications don't have local data folders at all, and for those cases we're adding special handling in the next release.

In your case, however, it would appear that the local data folder exists (that is, a valid path is returned from `Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData)`, but the application can't write to it.

Can you confirm that? If so, we'll add handling for that case as well.

Thank you!

**Comment by upsilondynamics** [comment]
Hey @ClearScriptLib ,

On my local machine (as in my development machine) everything works fine as expected, I am unable to confirm on Azure if:

**C:\DWASFiles\Sites\#1flowdynamics__048e\LocalAppData\Microsoft\ClearScript\6.0.0\x64**

Has write access because in the Kudo terminal I am only able to access a D:\ drive which is where my sites physical files are hosted the path above relative to the IIS instance I am unsure of.  But you can likely put in a check on a) If the directory exists and b) Is writeable then to use it.

My fix writes it to the same path as the site is, which I was able to verify works.  Could be some fallback condition you may be able to beefen up the logic.

**Comment by ClearScriptLib** [comment]
That seems like a reasonable approach, and it's similar to our upcoming fix for Issue #160. Thanks for your help!

**Comment by ClearScriptLib** [comment]
Fixed in [Version 6.0.1](https://github.com/microsoft/ClearScript/releases/tag/6.0.1).
